code/modules/atmospherics/ZAS/zas_extras/inflatable.dm

```diff
@@ -291,14 +296,105 @@
 			src.transfer_fingerprints_to(R)
 			qdel(src)
 
+//inflatable shelters from vg
+/obj/structure/inflatable/shelter
+	name = "inflatable shelter"
+	density = TRUE
+	anchored = FALSE
+	opacity = 0
+	can_buckle = TRUE
+	icon_state = "shelter_base"
+	undeploy_path = /obj/item/inflatable/shelter
+	var/list/exiting = list()
+	var/datum/gas_mixture/cabin_air
+
+/obj/structure/inflatable/shelter/New()
+	. = ..()
+	cabin_air = new
+	cabin_air.volume = CELL_VOLUME / 3
+	cabin_air.temperature = T20C + 20
+	cabin_air.adjustMultipleGases(
+		GAS_OXYGEN, MOLES_O2STANDARD,
+		GAS_NITROGEN, MOLES_N2STANDARD)
+
+/obj/structure/inflatable/shelter/examine(mob/user)
+	. = ..()
+	. += span_notice("Click to enter. Use grab on shelter to force target inside. Use resist to exit. Right click to deflate.")
+	var/list/living_contents = list()
+	for(var/mob/living/L in contents)
+		living_contents += L.name
+	if(living_contents.len)
+		. += span_notice("You can see [english_list(living_contents)] inside.")
+
+/obj/structure/inflatable/shelter/attack_hand(mob/user)
+	user.visible_message(span_notice("[user] begins to climb into \the [src]."), span_notice("You begin to climb into \the [src]."))
+	if(do_after(user, 3 SECONDS))
+		enter_shelter(user)
+
+/obj/structure/inflatable/shelter/user_buckle_mob(mob/living/M, mob/user, check_loc = TRUE)
+	if(!in_range(user, src) || !in_range(M, src))
+		return FALSE
+	if(do_after(user, 3 SECONDS))
+		if(!in_range(M, src))
+			return FALSE //in case the target has moved
+		enter_shelter(M)
+		return
+
+/obj/structure/inflatable/shelter/Destroy()
+	for(var/atom/movable/AM in src)
+		AM.forceMove(loc)
+	qdel(cabin_air)
+	cabin_air = null
+	exiting = null
+	..()
+
+/obj/structure/inflatable/shelter/remove_air(amount)
+	return cabin_air.remove(amount)
+
+/obj/structure/inflatable/shelter/return_air()
+	return cabin_air
+
+/obj/structure/inflatable/shelter/proc/enter_shelter(mob/user)
+	user.forceMove(src)
+	update_icon()
+	user.visible_message(span_notice("[user] enters \the [src]."), span_notice("You enter \the [src]."))
+
```

Comment:
Register a signal to the user for `COMSIG_PARENT_QDELETING`. Have the response proc remove it from the shelter's `vis_contents` to prevent harddels.

Unregister the signal when it exits.